gc/base/SparseAddressOrderedFixedSizeDataPool.hpp

```diff
@@ -0,0 +1,283 @@
+
+	/**
+	 * Get J9PortVmemIdentifier associated with data pointer
+	 *
+	 * @param dataPtr	void*		Data pointer
+	 * @return J9PortVmemIdentifier of data pointer
+	 */
+	struct J9PortVmemIdentifier* findIdentifierForSparseDataPtr(void *dataPtr);
+#endif /* OMR_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION */
+
+	/**
+	 * Add object entry to the hash table that maps the proxyObjPtr to the data pointer
+	 *
+	 * @param dataPtr		void*		data location pointer
+	 * @param proxyObjPtr	void*		Proxy object associated with dataPtr
+	 * @param size			uintptr_t	Size of region consumed by dataPtr
+	 *
+	 * @return true if object is added successfully to the hash table , false otherwise
+	 */
+	bool mapSparseDataPtrToHeapProxyObjectPtr(void *dataPtr, void *proxyObjPtr, uintptr_t size);
+
+	/**
+	 * Remove entry from the hash table that is associated the object data pointer provided
+	 *
+	 * @param dataPtr	void*	Data pointer
+	 *
+	 * @return true if key associated to dataPtr is removed successfully, false otherwise
+	 */
+	bool unmapSparseDataPtrFromHeapProxyObjectPtr(void *dataPtr);
+
+	/**
+	 * Get SparseDataTableEntry associated with data pointer
+	 *
+	 * @param dataPtr	void*	Data pointer
+	 * @return in-heap proxy object pointer of data pointer
+	 */
+	SparseDataTableEntry *findSparseDataTableEntryForSparseDataPtr(void *dataPtr);
+	/**
+	 * Get data size in bytes associated with the data pointer
+	 *
+	 * @param dataPtr	void*	Data pointer
+	 * @return size of data pointer in bytes
+	 */
+	uintptr_t findObjectDataSizeForSparseDataPtr(void *dataPtr);
+
+	/**
+	 * Get proxy object pointer associated with data pointer
+	 *
+	 * @param dataPtr	void*	Data pointer
+	 * @return in-heap proxy object pointer of data pointer
+	 */
+	void *findHeapProxyObjectPtrForSparseDataPtr(void *dataPtr);
+
+	/**
+	 * Check if the given data pointer is valid
+	 *
+	 * @param dataPtr	void*	Data pointer
+	 * @return true if data pointer is valid
+	 */
+	bool isValidDataPtr(void *dataPtr);
+
+	/**
+	 * Get the largest free list entry
+	 */
+	MMINLINE uintptr_t getLargestFreeEntry()
+	{
+		return _approxLargestFreeEntry;
+	}
+
+	/**
+	 * Set the largest free list entry
+	 */
+	MMINLINE void setLargestFreeEntry(uintptr_t approxLargestFreeEntry)
+	{
+		_approxLargestFreeEntry = approxLargestFreeEntry;
+	}
+
+	/**
+	 * Update the proxyObjPtr after an object has moved for the sparse data entry associated with the given dataPtr.
+	 *
+	 * @param dataPtr		void*	Data pointer
+	 * @param proxyObjPtr	void*	Updated in-heap proxy object pointer for data pointer
+	 *
+	 * @return true if the sparse data entry was successfully updated, false otherwise
+	 */
+	bool updateSparseDataEntryAfterObjectHasMoved(void *dataPtr, void *proxyObjPtr);
+
+	void printAllDebugInfo();
+
+protected:
+private:
+	bool initialize(MM_EnvironmentBase *env, void *sparseHeapBase);
+	void tearDown(MM_EnvironmentBase *env);
+	void kill(MM_EnvironmentBase *env);
```

Comment:
usually we declare `kill()` as `public` (as opposite to `newInstance()`) and `initialize()` and `tearDown()` as `protected` (to cover possible case in the future this class might have children need to call parent methods)

Reply:
Yes, good point. Will make these changes.